include/deal.II/matrix_free/cuda_fe_evaluation.cuh

```diff
@@ -0,0 +1,744 @@
+          __syncthreads();
+
+          values<1, true, false, true>(grad_u[0], grad_u[0]);
+          gradients<1, true, false, true>(grad_u[1], grad_u[1]);
+
+          break;
+        }
+        case 3:
+        {
+          gradients<0, true, false, false>(u, grad_u[0]);
+          values<0, true, false, false>(u, grad_u[1]);
+          values<0, true, false, false>(u, grad_u[2]);
+
+          __syncthreads();
+
+          values<1, true, false, true>(grad_u[0], grad_u[0]);
+          gradients<1, true, false, true>(grad_u[1], grad_u[1]);
+          values<1, true, false, true>(grad_u[2], grad_u[2]);
+
+          __syncthreads();
+
+          values<2, true, false, true>(grad_u[0], grad_u[0]);
+          values<2, true, false, true>(grad_u[1], grad_u[1]);
+          gradients<2, true, false, true>(grad_u[2], grad_u[2]);
+
+          break;
+        }
+        default:
+        {
+          // Do nothing. We should throw but we can't from a __device__ function.
+        }
+        }
+    }
+
+
+
+    template <int dim, int fe_degree, int n_q_points_1d, typename Number>
+    template <bool add>
+    inline
+    __device__ void EvaluatorTensorProduct<evaluate_general, dim, fe_degree,
+               n_q_points_1d, Number>::integrate_gradient(
+                 Number *u,
+                 Number *grad_u[dim])
+    {
+      switch (dim)
+        {
+        case 1:
+        {
+          gradients<0, false, add, false> (grad_u[dim], u);
+
+          break;
+        }
+        case 2:
+        {
+          gradients<0, false, false, true> (grad_u[0], grad_u[0]);
+          values<0, false, false, true> (grad_u[1], grad_u[1]);
+
+          __syncthreads();
+
+          values<1, false, add, false> (grad_u[0], u);
+          __syncthreads();
+          gradients<1, false, true, false> (grad_u[1], u);
+
+          break;
+        }
+        case 3:
+        {
+          gradients<0, false, false, true> (grad_u[0], grad_u[0]);
+          values<0, false, false, true> (grad_u[1], grad_u[1]);
+          values<0, false, false, true> (grad_u[2], grad_u[2]);
+
+          __syncthreads();
+
+          values<1, false, false, true> (grad_u[0], grad_u[0]);
+          gradients<1, false, false, true> (grad_u[1], grad_u[1]);
+          values<1, false, false, true> (grad_u[2], grad_u[2]);
+
+          __syncthreads();
+
+          values<2, false, add, false> (grad_u[0], u);
+          __syncthreads();
+          values<2, false, true, false> (grad_u[1], u);
+          __syncthreads();
+          gradients<2, false, true, false> (grad_u[2], u);
+
+          break;
+        }
+        default:
+        {
+          // Do nothing. We should throw but we can't from a __device__ function.
+        }
+        }
+    }
+
+
+
+    template <int dim, int fe_degree, bool transpose, typename Number>
+    __device__ void resolve_hanging_nodes_shmem(Number *values, const unsigned
+                                                int constr)
+    {
+      //TODO
+    }
+  }
```

Comment:
In the CPU code, we moved these functions to a file `tensor_product_kernels.h`. The GPU `FEEvaluation` file isn't that large yet but I would still suggest to use a similar organization.